[SPARK-16992][PYSPARK] Python Pep8 formatting and import reorganisation

gsemet opened the pull request:
## What changes were proposed in this pull request?

This patch adds a code style validation script named `dev/py-validate.sh` to enforce pep8 recommendations.

The execution of this script will format a lot of files. I did not put them into this PR. I plan to execute this smoothly and reformat the code in various Pull Requests to ease reviews.

First one is automatic formatting of the examples in the documentations: See #14830.

Features of the current patch:
- add a `python/.editconfig` file (Scala files use 2 space indentation, while Python files uses 4) for compatible editors (almost every editors has a plugin to support .editconfig file)
- use `autopep8` to fix basic pep8 mistakes
- use `isort` to automatically split "import" statement and organise them into logically linked order (see doc [here](https://pypi.python.org/pypi/isort#multi-line-output-modes). The most important thing is that it split import statements that imports more than one object into several lines. This will increase the number of line of the file, but this facilitates a lot file maintainance and file merges if needed.
- add a `py-validate.sh` script in order to automatise the correction (need isort and autopep8 installed inside a virtualenv that has the right version of the tools, this is documented in the script)

You can see similar script in prod in the [Buildbot](https://github.com/buildbot/buildbot/blob/master/common/validate.sh) project
## How was this patch tested?

Simple tests on my machines has been done (local mode only). There should not have any regression or feature change at all with this pull request.

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented:
Although there has generally been some resistance to large style-only changes, we do enforce import order in Scala/Java including checks. So it seems pretty reasonable to do the same in one big go for Python.

srowen commented on python/pyspark/context.py:
```diff
@@ -22,22 +22,30 @@
+from pyspark.rdd import RDD
+from pyspark.rdd import _load_from_socket
+from pyspark.rdd import ignore_unicode_prefix
+from pyspark.serializers import AutoBatchedSerializer
```
Expanding these multiple imports seems counter-productive. We don't do it in Scala (and in Java you can only import one thing or everything). Is this important/canonical for PEP8?

Stibbons replied:
indeed this is a deviation from Pep recommendation. I encourage this behavior since it simplifies a lot file maintainance.

On our Buildbot based project, we use to have lot of conflict involving changes on import statement: on 2 differents branches (say: prod and main), we often had to import either the same import (so merge might add this line twice when the two developers have placed them in two different places) or not so easy to solve conflict (when two developers add theo import two different object from the same module).
Once we setup the "one import per line" rule, no more conflict on this lines, ever. This helped us a lot automatizing an auto merger from the release branches to the "master" branch.

srowen replied:
Yeah we get that in the Scala code for sure, although you get conflicts even if you change merely adjacent lines anyway, so it only saves so much conflict. It's a decent open question, I wonder what others think? the downside is the extra duplication of import boilerplate.

Stibbons replied:
As you can see in https://pypi.python.org/pypi/isort, there a several way to format this multiple import statements. At worst, at least I recommende to enforce the sort of this multi import lines so there is no ambiguity where to place any "import" (and isort will correct the change from the developer)

Stibbons commented:
Rebased, sorry I had to force push this PR.

Stibbons commented on python/run-tests.py:
```diff
@@ -37,11 +45,6 @@
 
-from sparktestsupport import SPARK_HOME  # noqa (suppress pep8 warnings)
-from sparktestsupport.shellutils import which, subprocess_check_output  # noqa
-from sparktestsupport.modules import all_modules  # noqa
```
is it a problem if these statements are placed higher in the file?

rxin commented:
Can you create a JIRA ticket for this? This is too large to go in without a JIRA ticket.

rxin commented:
BTW this is actually a non-trivial change and would require very careful look, since Python imports are not side effect free.

srowen commented:
@Stibbons connect this to your JIRA in the title
https://cwiki.apache.org/confluence/display/SPARK/Contributing+to+Spark

Stibbons commented:
Done

srowen commented:
Jenkins test this please

srowen commented:
Jenkins add to whitelist

SparkQA commented:
**[Test build #63533 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63533/consoleFull)** for PR 14567 at commit [`067ab4a`](https://github.com/apache/spark/commit/067ab4a3c5ca5a42be87a47b0bcb17870465c130).

SparkQA commented:
**[Test build #63533 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63533/consoleFull)** for PR 14567 at commit [`067ab4a`](https://github.com/apache/spark/commit/067ab4a3c5ca5a42be87a47b0bcb17870465c130).
 * This patch **fails RAT tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63533/
Test FAILed.

SparkQA commented:
**[Test build #63534 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63534/consoleFull)** for PR 14567 at commit [`6f17382`](https://github.com/apache/spark/commit/6f173821789c1b2b371cd4c368c4b9b643c8eef5).

SparkQA commented:
**[Test build #63534 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63534/consoleFull)** for PR 14567 at commit [`6f17382`](https://github.com/apache/spark/commit/6f173821789c1b2b371cd4c368c4b9b643c8eef5).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63534/
Test FAILed.

SparkQA commented:
**[Test build #63536 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63536/consoleFull)** for PR 14567 at commit [`239344b`](https://github.com/apache/spark/commit/239344b2bb833ec795a85a966baa0169d052b8b5).

SparkQA commented:
**[Test build #63536 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63536/consoleFull)** for PR 14567 at commit [`239344b`](https://github.com/apache/spark/commit/239344b2bb833ec795a85a966baa0169d052b8b5).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63536/
Test FAILed.

SparkQA commented:
**[Test build #63544 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63544/consoleFull)** for PR 14567 at commit [`b3a176c`](https://github.com/apache/spark/commit/b3a176c53b16cfcd2d9d796036284c1fb18cf951).

SparkQA commented:
**[Test build #63544 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63544/consoleFull)** for PR 14567 at commit [`b3a176c`](https://github.com/apache/spark/commit/b3a176c53b16cfcd2d9d796036284c1fb18cf951).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63544/
Test PASSed.

holdenk commented on python/pep8rc:
```diff
@@ -0,0 +1,21 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+[pep8]
```
There is another pep8config file at ./dev/toxi.ini - seems like it would be good to have a single file (also unify the ignore lists)

Stibbons replied:
done

srowen replied:
Is the rename of tox.ini addressing this -- I wasn't sure how

srowen replied:
Maybe I'm missing something but should they be in the same file or are they separate?

Stibbons replied:
I don't know if they can be merged. Will try it

srowen replied:
I'm ignorant of the config here of course, but it seems like we've replaced 1 config file (tox.ini) with 4 in total (pep8rc, isort.cfg, style.yapf.ini, .editorconfig). Maybe it would help to explain what the differences are, because maybe only some of them are really worth dealing with.

Stibbons replied:
- tox.ini kind of becames this pep8rc, indeed.
- isort only deals with sorting "import" statement, not checked by pep8, but pylint has some checks on it
- style.yapf.ini is the configuration for yapf, which is a more aggressive version of 'autopep8' made by google for their chrominum projects, to strictly format all their python files. I don't recommend to enforce the usage of it, but if one want to use it on its python file in the Spark project, he can execute it and review the style before commiting. In an ideal wold, this would be a mandatory formatting (on post commit hook for example), but this may take some times before reaching this point
- .editorconfig only deals with tab space, like I said, might be put outside of this PR if you want

holdenk commented on python/pyspark/cloudpickle.py:
```diff
@@ -194,7 +194,7 @@ def save_function(self, obj, name=None):
         # reference (as is done in default pickler), via save_function_tuple.
         if islambda(obj) or obj.__code__.co_filename == '<stdin>' or themodule is None:
-            #print("save global", islambda(obj), obj.__code__.co_filename, modname, themodule)
+            # print("save global", islambda(obj), obj.__code__.co_filename, modname, themodule)
```
Seems like we might just want to remove this commented out line?

Stibbons replied:
here are the list of ignored files in tox.ini: cloudpickle.py,heapq3.py,shared.py
Look like it would be better to style them anyway. Could I remove them from the exclusion list?

SparkQA commented:
**[Test build #63689 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63689/consoleFull)** for PR 14567 at commit [`ff549b6`](https://github.com/apache/spark/commit/ff549b6272cf31ed1445ab6733d8314d3af336a3).

SparkQA commented:
**[Test build #63689 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63689/consoleFull)** for PR 14567 at commit [`ff549b6`](https://github.com/apache/spark/commit/ff549b6272cf31ed1445ab6733d8314d3af336a3).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63689/
Test FAILed.

SparkQA commented:
**[Test build #63690 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63690/consoleFull)** for PR 14567 at commit [`be39a83`](https://github.com/apache/spark/commit/be39a83e3cd381b8c9cc72b5ca202693257c07e7).

SparkQA commented:
**[Test build #63690 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63690/consoleFull)** for PR 14567 at commit [`be39a83`](https://github.com/apache/spark/commit/be39a83e3cd381b8c9cc72b5ca202693257c07e7).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63690/
Test FAILed.

SparkQA commented:
**[Test build #63691 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63691/consoleFull)** for PR 14567 at commit [`46e83f2`](https://github.com/apache/spark/commit/46e83f241f9a73e3bc42fa72eb458ee98de9ebed).

SparkQA commented:
**[Test build #63691 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63691/consoleFull)** for PR 14567 at commit [`46e83f2`](https://github.com/apache/spark/commit/46e83f241f9a73e3bc42fa72eb458ee98de9ebed).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63691/
Test FAILed.

SparkQA commented:
**[Test build #63693 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63693/consoleFull)** for PR 14567 at commit [`d3eab9f`](https://github.com/apache/spark/commit/d3eab9f5abacbd4112bf7e5ab78ce069edb02a7d).

SparkQA commented:
**[Test build #63693 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63693/consoleFull)** for PR 14567 at commit [`d3eab9f`](https://github.com/apache/spark/commit/d3eab9f5abacbd4112bf7e5ab78ce069edb02a7d).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63693/
Test FAILed.

SparkQA commented:
**[Test build #63694 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63694/consoleFull)** for PR 14567 at commit [`500b659`](https://github.com/apache/spark/commit/500b659341ab2bda57137ca234ad7b63ae71822b).

Stibbons commented:
I reworked my Pull Request:
- tox.ini and pep8rc are factorized
- manual pep8 fixes that are not done by autopep8 are done
- there is NO automatic formatting of the code. I propose to split this job into 2 PR, one that is for script, config, and manual pep8 fixes, and one with only automatic pep8/isort fixes.I will submit as a new pull request once this one has been accepted. For information, it changes more than 1300 loc, so for easing the review I have splitted. To find out the final state of each file, here is the [full diff for information](https://github.com/apache/spark/compare/master...Stibbons:python_import_reorg_plus_exec?expand=1) 

I have switched to use the "Vertical" policy on multiple package import. See how this [file will be formatted](https://github.com/apache/spark/compare/master...Stibbons:python_import_reorg_plus_exec?expand=1#diff-102). When if fit on 1 line, the module are just sorted. When it does not fit on 1 line, we use the "(" syntax for import.

Let me know what you think of this PR.

Stibbons commented on python/pyspark/cloudpickle.py:
```diff
@@ -42,17 +42,17 @@
 """
```
isort/pep8 executed on this file to make jenkins happy

davies replied:
This file is copied, please ignore it

Stibbons commented on python/pyspark/ml/param/shared.py:
```diff
@@ -25,7 +25,8 @@ class HasMaxIter(Params):
     Mixin for param maxIter: max number of iterations (>= 0).
```
isort/pep8 executed on this file to make jenkins happy

davies replied:
This file is generated by _shared_params_code_gen.py., can we ignore it?

Stibbons replied:
i got error in jenkins if I don't modify it

SparkQA commented:
**[Test build #63694 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63694/consoleFull)** for PR 14567 at commit [`500b659`](https://github.com/apache/spark/commit/500b659341ab2bda57137ca234ad7b63ae71822b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63694/
Test PASSed.

Stibbons commented:
Any update?

davies commented on python/pyspark/heapq3.py:
```diff
@@ -408,10 +408,12 @@
 __all__ = ['heappush', 'heappop', 'heapify', 'heapreplace', 'merge',
            'nlargest', 'nsmallest', 'heappushpop']
 
+
```
This file is copied, please ignore it

Stibbons replied:
Done

davies commented on dev/py-validate.sh:
```diff
@@ -0,0 +1,110 @@
+#!/usr/bin/env bash
```
Where is this file used?

Stibbons replied:
it is the script that format all python code automatically

srowen replied:
@davies what do you think about including this vs just enforcing the checks? If it would be used, I'm for it, but not clear how often this would be run in practice by anyone if we're not going to use its big-bang changes.

Stibbons replied:
My point of view:
- don't enforce it right away, just agree on the style
- execute manualy on some files (in my next PR, it's on documentation), and on small big bang PRs
- then enforce automatic formatting (preferably with Yapf) on some files. On Buildbot we use a git commit hook to check python files automatically, apply some pep8 formatting, execute pylint,...

Stibbons commented on python/pyspark/cloudpickle.py:
```diff
@@ -280,7 +279,7 @@ def extract_code_globals(co):
         if co.co_consts:
             for const in co.co_consts:
-                if type(const) is types.CodeType:
+                if isinstance(const, types.CodeType):
```
this is quite a big error, is it enough to fix it here?

srowen replied:
Hm, i think this is third-party code. I'm reluctant to change it because it's just a straight copy of the file from a particular release. Can we exclude it from pep8?

You're welcome of course to file a bug against cloudpickle.

Stibbons replied:
this file is now ignored actually

Stibbons commented:
Reworked according to your review. Please keep in mind this is just the first part of a two part pull request, the second will contain a selected part of [this diff](https://github.com/apache/spark/compare/master...Stibbons:python_import_reorg_plus_exec?expand=1), which is the result after the execution of ```py-validate.py```.

SparkQA commented:
**[Test build #64189 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64189/consoleFull)** for PR 14567 at commit [`015028f`](https://github.com/apache/spark/commit/015028f6647c13bcd589133b70b79049b0a479aa).

SparkQA commented:
**[Test build #64190 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64190/consoleFull)** for PR 14567 at commit [`0801573`](https://github.com/apache/spark/commit/0801573c57b701c9d59e8eb311ae31b42a1c7267).

SparkQA commented:
**[Test build #64189 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64189/consoleFull)** for PR 14567 at commit [`015028f`](https://github.com/apache/spark/commit/015028f6647c13bcd589133b70b79049b0a479aa).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64189/
Test FAILed.

SparkQA commented:
**[Test build #64190 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64190/consoleFull)** for PR 14567 at commit [`0801573`](https://github.com/apache/spark/commit/0801573c57b701c9d59e8eb311ae31b42a1c7267).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64190/
Test PASSed.

Stibbons commented:
@davies, @holdenk,@srowen, @rxin do you have time review this PR ? thx

holdenk commented:
@Stibbons I'm on vacation right now and I'm trying to take a bit of a break from working on Spark directly (mostly to catch up on my Spark related writing :p) - but I can take a look at it over the weekend if that works with you.

holdenk commented:
Will steed need one of the committers to take another look at some point of course :)

srowen commented on python/run-tests.py:
```diff
@@ -37,9 +37,9 @@
-from sparktestsupport import SPARK_HOME  # noqa (suppress pep8 warnings)
-from sparktestsupport.shellutils import which, subprocess_check_output  # noqa
-from sparktestsupport.modules import all_modules  # noqa
+from sparktestsupport import SPARK_HOME  # noqa (suppress pep8 warnings) isort:skip
```
For my info, what does isort:skip do?

Stibbons replied:
it prevent these lines to be automatically moved on top of the sys.path.append. For me this is weird to have to modify the sys.path, probably because pyspark unit test are not executed inside a virtualenv

srowen commented on .isort.cfg:
```diff
@@ -1,9 +1,9 @@
 # Licensed to the Apache Software Foundation (ASF) under one or more
-# contributor license agreements.  See the NOTICE file distributed with
+# contributor license agreements. See the NOTICE file distributed with
```
I don't think we want to 'fix' the standard license header necessarily unless it's really hard to exclude. Is this failing PEP8?

Also does this now put a file at the project root? I spent time getting stuff out of the root and would prefer to keep everything under python/

Stibbons replied:
With latest version it seenms I can put it in the ```dev``` folder

srowen commented on .editorconfig:
```diff
@@ -0,0 +1,30 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
```
Same, would not want to put tis in the project root.

Stibbons replied:
Should be at the project root, for editors to find it. The spec of all plugin is set that way: plugins will search from the opened file and go to parent dirs until it find a .editorconfig. See spec at http://editorconfig.org/

Most plugins does not support specifying an hardcoded path (like ```./dev```). For example the sublime editorconfig does not support it.

srowen replied:
Would this still work under python/ because that's where Python-related scripts would be run from?

Stibbons replied:
yes of course

Question: do you want to apply this fix on other python scripts (for example, the python code from the documentation) ?

srowen replied:
Meaning, do we need to fix up the style on other Python code to match what this pep8 config would enforce? yes ideally so. Much of the code in the docs are actually broken out as source files that are inlined, so hopefully they can be tested directly by pep8. For stuff that isn't otherwise tested, yes, making it match the enforced style is worthwhile.

Stibbons commented:
thanks a lot :)

SparkQA commented:
**[Test build #64469 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64469/consoleFull)** for PR 14567 at commit [`9696ea2`](https://github.com/apache/spark/commit/9696ea2a5e7f09f563d97f9916bcac81aee37161).

SparkQA commented:
**[Test build #64470 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64470/consoleFull)** for PR 14567 at commit [`92bb87e`](https://github.com/apache/spark/commit/92bb87e2b217944236f8fd6ac74cac3e087a2851).

Stibbons commented:
changes:
- recommendation to run py-validate.sh from a virtualenv with the versions of pep8, autopep8 and isort fixed (set in dev/requirements.txt)
- isort.cfg and .editorconfig no more on root

I propose to commit the modified files in the future in small pull requests for easy review

SparkQA commented:
**[Test build #64475 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64475/consoleFull)** for PR 14567 at commit [`ba3a6d1`](https://github.com/apache/spark/commit/ba3a6d14a083e77a04bf2ea5e38fc4f7b841af56).

SparkQA commented:
**[Test build #64477 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64477/consoleFull)** for PR 14567 at commit [`975dd17`](https://github.com/apache/spark/commit/975dd17a5a530f41c94d713ce005bb0be6d84d12).

Stibbons commented:
changes:

- use pycodestyle (replaces 'pep8' tool since may 2016)
- keep using autopep8 for light modifications
- Add yapf style ```dev/style.yapf.ini``` to people that wish to use it
  (it is really aggressive in formatting)

SparkQA commented:
**[Test build #64477 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64477/consoleFull)** for PR 14567 at commit [`975dd17`](https://github.com/apache/spark/commit/975dd17a5a530f41c94d713ce005bb0be6d84d12).
 * This patch **fails RAT tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64477/
Test FAILed.

SparkQA commented:
**[Test build #64478 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64478/consoleFull)** for PR 14567 at commit [`5b16a96`](https://github.com/apache/spark/commit/5b16a96e862360dcf223bc77fd791727394d346c).

SparkQA commented:
**[Test build #64469 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64469/consoleFull)** for PR 14567 at commit [`9696ea2`](https://github.com/apache/spark/commit/9696ea2a5e7f09f563d97f9916bcac81aee37161).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64469/
Test FAILed.

SparkQA commented:
**[Test build #64470 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64470/consoleFull)** for PR 14567 at commit [`92bb87e`](https://github.com/apache/spark/commit/92bb87e2b217944236f8fd6ac74cac3e087a2851).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64470/
Test PASSed.

SparkQA commented:
**[Test build #64475 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64475/consoleFull)** for PR 14567 at commit [`ba3a6d1`](https://github.com/apache/spark/commit/ba3a6d14a083e77a04bf2ea5e38fc4f7b841af56).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64475/
Test PASSed.

SparkQA commented:
**[Test build #64478 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64478/consoleFull)** for PR 14567 at commit [`5b16a96`](https://github.com/apache/spark/commit/5b16a96e862360dcf223bc77fd791727394d346c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64478/
Test PASSed.

srowen commented on dev/isort.cfg:
```diff
@@ -1,9 +1,9 @@
 # Licensed to the Apache Software Foundation (ASF) under one or more
-# contributor license agreements.  See the NOTICE file distributed with
+# contributor license agreements. See the NOTICE file distributed with
```
Does PEP8 really pick this up as a violation? what's the problem?

srowen replied:
Also could you comment on renaming tox.ini, I wasn't sure was requiring that

Stibbons replied:
No, this is not pep8 related!

I deleted dev/tox.ini and created a new file dev/isort.cfg from scratch. But the fact is git "finds" file renames based on content similarities, so it sees this as a file rename. Actually, I did deleted the tox.ini in another commit than the commit that creates isort.ini (see https://github.com/apache/spark/pull/14567/commits), so I guess this "visualisation" of a file renaming comes from github

srowen replied:
Hm, really, it guesses at moves? OK, if the net effect is that `tox.ini` has been deleted, OK. If you wouldn't mind, if you make another pass here, just add back those spaces. They seem to be present in the 'canonical' text of the license and best to just avoid any minor deviation regardless of where it came from.

So the `lint-python` script still works with this I take it because I see you updated it, and that's the important thing.

Stibbons replied:
yes it still works

srowen commented on python/.editorconfig:
```diff
@@ -0,0 +1,30 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+root = true
```
OK, just confirming that there are 2 additional config files needed to fully specify the style rules to enforce? what do these additional configs support beyond PEP8?

Stibbons replied:
This file is to automatically configure editors that have the "editorconfig" plugin to configure 2 spaces tab for scala files and 4 spaces tab on Python files. It is not related to pep8, so can be submitted in another PR if you prefer.

That's why I would prefer to be placed on the root of the project (and not the root of the python folder)

srowen commented:
OK I'd leave these changes to Python people like @davies @holdenk @MLnick to comment on from here. I think style changes can be OK if they're consistent, enforceable, and moving the code towards more standard style.

Stibbons commented on .gitignore:
```diff
@@ -28,6 +28,7 @@ build/*.jar
 build/apache-maven*
 build/scala*
 build/zinc*
+build/venv
```
I am used to put sphinx pep8, ... etc inside a virtual env here

Stibbons commented:
Rebased.

Update: 
- move ```.editorconfig``` up to the root of the project. This is needed so editors plugin will find it and configure both scala and python files. I didn't found a way to put it into the ```dev``` folder. Other possibility is to duplicate it (one in ```dev```, one in ```python```, one in ```mesos```, and so on ... not very elegant)
-

SparkQA commented:
**[Test build #64785 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64785/consoleFull)** for PR 14567 at commit [`656eb2e`](https://github.com/apache/spark/commit/656eb2e6ffe6f3161a7e735d4889f0206696bb05).

SparkQA commented:
**[Test build #64785 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64785/consoleFull)** for PR 14567 at commit [`656eb2e`](https://github.com/apache/spark/commit/656eb2e6ffe6f3161a7e735d4889f0206696bb05).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64785/
Test PASSed.

SparkQA commented:
**[Test build #64942 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64942/consoleFull)** for PR 14567 at commit [`1a2aa35`](https://github.com/apache/spark/commit/1a2aa35b7518ca08690381245b0636302f6e971f).

holdenk commented:
Hi @Stibbons maybe you can break down a bit more what the relationship between your two PRs and update the descriptions? I'm maybe a getting a bit fuzzy on what is intended to go where.

I like splitting up the PRs especially when they become this big - but I'm thinking if your going to do a huge re-org in the next PR it would make sense to have that PR enable full tests for that directory - perhaps as part of lint-python or similar mechanism so that as we incrementally fix sub-components we can ensure we don't move backwards.

Stibbons commented:
I am not sure if there is a test in pylint on the backslash syntax, there are some cases like with the ```with``` statement where the backslash might not be easily replaceable (see https://www.python.org/dev/peps/pep-0008/#maximum-line-length).

I update the description

SparkQA commented:
**[Test build #64942 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64942/consoleFull)** for PR 14567 at commit [`1a2aa35`](https://github.com/apache/spark/commit/1a2aa35b7518ca08690381245b0636302f6e971f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64942/
Test PASSed.

SparkQA commented:
**[Test build #65762 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65762/consoleFull)** for PR 14567 at commit [`b81c38f`](https://github.com/apache/spark/commit/b81c38f28799c72f65383c2d82904e1b92a8d7ec).

Stibbons commented:
Rebased

SparkQA commented:
**[Test build #65762 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65762/consoleFull)** for PR 14567 at commit [`b81c38f`](https://github.com/apache/spark/commit/b81c38f28799c72f65383c2d82904e1b92a8d7ec).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65762/
Test PASSed.

SparkQA commented:
**[Test build #66043 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66043/consoleFull)** for PR 14567 at commit [`a02f634`](https://github.com/apache/spark/commit/a02f634dca4975e0cd8656e865270ca7a3a8a4be).

SparkQA commented:
**[Test build #66043 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66043/consoleFull)** for PR 14567 at commit [`a02f634`](https://github.com/apache/spark/commit/a02f634dca4975e0cd8656e865270ca7a3a8a4be).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66043/
Test PASSed.

SparkQA commented:
**[Test build #66389 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66389/consoleFull)** for PR 14567 at commit [`4a3cde7`](https://github.com/apache/spark/commit/4a3cde78f52565d9ea6dee0cc34540702a2ce1fb).

SparkQA commented:
**[Test build #66389 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66389/consoleFull)** for PR 14567 at commit [`4a3cde7`](https://github.com/apache/spark/commit/4a3cde78f52565d9ea6dee0cc34540702a2ce1fb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66389/
Test PASSed.

gsemet committed:
[SPARK-16992][PYSPARK] Reenable Pylint

Use a virtualenv for isolation and easy installation.

This basically reverts 85a50a6352b72c4619d010e29e3a76774dbc0c71

Might have been a solution to SPARK-9385.

Lot of new test disabled. I propose to fix issues in various
pull requests (obviously most 'import' order errors should be
fixed by my other pull requests such as #14830 for documentation
examples, which is part of the effort on code style described
in #14567). Each subsequent pull request will fix one or more
error and reenable the according pylint check.

List of new disabled checks:
- bad-super-call
- consider-iterating-dictionary
- consider-using-enumerate
- eval-used
- exec-used
- invalid-length-returned
- misplaced-comparison-constant
- raising-bad-type
- redefined-variable-type
- trailing-newlines
- trailing-whitespace
- ungrouped-imports
- unnecessary-pass
- unneeded-not
- wrong-import-order
- wrong-import-position

Signed-off-by: Gaetan Semet <gaetan@xeberon.net>

gsemet committed:
better usage of debug mode

gsemet committed:
Leave anaconda environment in lint-pylint

gsemet committed:
new pylint ignore after rebase

+ quiet mode for sphinx

SparkQA commented:
**[Test build #66847 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66847/consoleFull)** for PR 14567 at commit [`1252d08`](https://github.com/apache/spark/commit/1252d0863fc1901aa8b87a5472fb42d6058b9600).

SparkQA commented:
**[Test build #66847 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66847/consoleFull)** for PR 14567 at commit [`1252d08`](https://github.com/apache/spark/commit/1252d0863fc1901aa8b87a5472fb42d6058b9600).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66847/
Test PASSed.

ueshin commented:
Hi, are you still working on this?
